### PR TITLE
[InstCombine] Fix Failure to fold (and %x, (sext i1 %m)) -> (select %m, %x, 0) with multiple uses of %m

### LINKED ISSUE #81288: [InstCombine] Failure to fold (and %x, (sext i1 %m)) -> (select %m, %x, 0) with multiple uses of %m
Godbolt: https://simd.godbolt.org/z/ndz15vGjW

```ll
define <2 x i64> @src(<2 x i64> %x, <2 x i64> %y, <2 x i64> %a, <2 x i64> %b) {
entry:
  %cmp.i8 = icmp sgt <2 x i64> %x, %y
  %sext.i9 = sext <2 x i1> %cmp.i8 to <2 x i64>
  %and.i7 = and <2 x i64> %sext.i9, %a
  %and.i = and <2 x i64> %sext.i9, %b
  %add.i = add <2 x i64> %and.i7, %and.i
  ret <2 x i64> %add.i
}

define <2 x i64> @tgt(<2 x i64> %x, <2 x i64> %y, <2 x i64> %a, <2 x i64> %b) {
entry:
  %0 = icmp sgt <2 x i64> %x, %y
  %1 = add <2 x i64> %b, %a
  %add.i = select <2 x i1> %0, <2 x i64> %1, <2 x i64> zeroinitializer
  ret <2 x i64> %add.i
}
```

The tgt version gives better codegen:
```asm
src:                                    # @src
        pcmpgtq %xmm1, %xmm0
        pand    %xmm0, %xmm2
        pand    %xmm3, %xmm0
        paddq   %xmm2, %xmm0
        retq
tgt:                                    # @tgt
        pcmpgtq %xmm1, %xmm0
        paddq   %xmm3, %xmm2
        pand    %xmm2, %xmm0
        retq
```

I can try and address this in DAG but middle end seems to be the better choice.

### PR DESCRIPTION
**Pull request by SahilPatidar**
Resolve #81288

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/and.ll [hunk 1]
```diff
@@ -2768,3 +2768,18 @@ define i32 @add_constant_equal_with_the_top_bit_of_demandedbits_insertpt(i32 %x,
...
+; CHECK-NEXT:    [[ADD:%.*]] = select i1 [[CMP]], i32 [[TMP1]], i32 0
+; CHECK-NEXT:    ret i32 [[ADD]]
+;
+  %cmp = icmp sgt i32 %x, %y
+  %sext = sext i1 %cmp to i32
+  %and1 = and i32 %sext, %a
+  %and2 = and i32 %sext, %b
+  %add = add i32 %and1, %and2
+  ret i32 %add
+}
```
Comment (RKSimon) [inline llvm/test/Transforms/InstCombine/and.ll:2785]: You need a better name - and_sext_multiuse at the very least. You also need to add some variants of the test with commuted operands / different types of use of the sext - maybe see what happens if you temporarily disable the and(sext(A), B) fold entirely and use the update script to see what other tests are dependent on this fold and what form they take.

**Comment by RKSimon** [comment]
@nikic are you happy with the test coverage here?

File: llvm/test/Transforms/InstCombine/and.ll [hunk 2]
```diff
@@ -1818,7 +1818,7 @@ define i8 @not_ashr_bitwidth_mask_use2(i8 %x, i8 %y) {
 ; CHECK-NEXT:    [[ISNOTNEG:%.*]] = icmp sgt i8 [[X:%.*]], -1
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/and.ll:1818]: Adjust the comment above saying "negative test - extra use". It's no longer a negative test.

File: llvm/test/Transforms/InstCombine/and.ll [hunk 3]
```diff
@@ -1925,7 +1925,7 @@ define i16 @invert_signbit_splat_mask_use3(i8 %x, i16 %y) {
...
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/and.ll:1925]: Same here.

File: llvm/test/Transforms/InstCombine/and.ll (hunk 2 above)
Comment (SahilPatidar) [reply llvm/test/Transforms/InstCombine/and.ll:1818]: Let me know if it is not correct?

**Review by nikic — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
Actually, it looks like there are two legitimate test failures:
```
Failed Tests (2):
  LLVM :: Transforms/InstCombine/binop-cast.ll
  LLVM :: Transforms/InstCombine/sub-ashr-and-to-icmp-select.ll
```
Can you please update these tests as well?

**Comment by SahilPatidar** [comment]
Please let me know if any changes need to be made.

File: llvm/test/Transforms/InstCombine/binop-cast.ll [hunk 4]
```diff
@@ -50,7 +50,7 @@ define i32 @and_sext_to_sel_multi_use(i32 %x, i1 %y) {
 ; CHECK-LABEL: @and_sext_to_sel_multi_use(
 ; CHECK-NEXT:    [[SEXT:%.*]] = sext i1 [[Y:%.*]] to i32
 ; CHECK-NEXT:    call void @use(i32 [[SEXT]])
-; CHECK-NEXT:    [[R:%.*]] = and i32 [[SEXT]], [[X:%.*]]
+; CHECK-NEXT:    [[R:%.*]] = select i1 [[Y]], i32 [[X:%.*]], i32 0
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/binop-cast.ll:53]: Think this one could classified as regression but probably not a big deal.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**